numba/unicode.py

```diff
@@ -1582,110 +1955,118 @@ def impl(a):
             return ret
     return impl
 
+# generator for simple unicode "isX" methods
 
-# https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L11896-L11925    # noqa: E501
-@overload_method(types.UnicodeType, 'isspace')
-def unicode_isspace(data):
-    """Implements UnicodeType.isspace()"""
-
-    def impl(data):
-        length = len(data)
-        if length == 1:
-            return _PyUnicode_IsSpace(_get_code_point(data, 0))
 
-        if length == 0:
-            return False
+def gen_isX(_PyUnicode_IS_func, empty_is_false=True):
+    def unicode_isX(data):
+        def impl(data):
+            length = len(data)
+            if length == 1:
+                return _PyUnicode_IS_func(_get_code_point(data, 0))
 
-        for i in range(length):
-            code_point = _get_code_point(data, i)
-            if not _PyUnicode_IsSpace(code_point):
+            if empty_is_false and length == 0:
                 return False
 
-        return True
+            for i in range(length):
+                code_point = _get_code_point(data, i)
+                if not _PyUnicode_IS_func(code_point):
+                    return False
 
-    return impl
+            return True
 
+        return impl
+    return unicode_isX
 
-@overload_method(types.UnicodeType, 'istitle')
-def unicode_istitle(s):
-    """
-    Implements UnicodeType.istitle()
-    The algorithm is an approximate translation from CPython:
-    https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L11829-L11885 # noqa: E501
-    """
 
-    def impl(s):
-        cased = False
-        previous_is_cased = False
-        for char in s:
-            if _PyUnicode_IsUppercase(char) or _PyUnicode_IsTitlecase(char):
-                if previous_is_cased:
-                    return False
-                cased = True
-                previous_is_cased = True
-            elif _PyUnicode_IsLowercase(char):
-                if not previous_is_cased:
-                    return False
-            else:
-                previous_is_cased = False
+# https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L11896-L11925    # noqa: E501
+overload_method(types.UnicodeType, 'isspace')(gen_isX(_PyUnicode_IsSpace))
 
-        return cased
-    return impl
+# https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L12096-L12124    # noqa: E501
+overload_method(types.UnicodeType, 'isnumeric')(gen_isX(_PyUnicode_IsNumeric))
 
+# https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L12056-L12085    # noqa: E501
+overload_method(types.UnicodeType, 'isdigit')(gen_isX(_PyUnicode_IsDigit))
+
+# https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L12017-L12045    # noqa: E501
+overload_method(
```

Comment:
The formatting here is awkward. perhaps something like the below is better:

```python
overload_method(types.UnicodeType, 'isdecimal')(
    gen_isX(_PyUnicode_IsDecimalDigit),
)
```

Reply:
I have to confess to just running an auto-formatter after doing all the merging/resolving conflicts. I've fixed this in d2584f6.